File: api/src/main/scala/com/github/pflooky/datacaterer/api/DataCatererConfigurationBuilder.scala
package com.github.pflooky.datacaterer.api

import com.github.pflooky.datacaterer.api.connection.{CassandraBuilder, ConnectionTaskBuilder, FileBuilder, HttpBuilder, KafkaBuilder, MySqlBuilder, PostgresBuilder, SolaceBuilder}
import com.github.pflooky.datacaterer.api.converter.Converters.toScalaMap
import com.github.pflooky.datacaterer.api.model.Constants._
import com.github.pflooky.datacaterer.api.model.DataCatererConfiguration
import com.softwaremill.quicklens.ModifyPimp

case class DataCatererConfigurationBuilder(build: DataCatererConfiguration = DataCatererConfiguration()) {
  def this() = this(DataCatererConfiguration())

  def master(master: String): DataCatererConfigurationBuilder =
    this.modify(_.build.master).setTo(master)

  def runtimeConfig(conf: Map[String, String]): DataCatererConfigurationBuilder =
    this.modify(_.build.runtimeConfig)(_ ++ conf)

  def runtimeConfig(conf: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    runtimeConfig(toScalaMap(conf))

  def addRuntimeConfig(conf: (String, String)): DataCatererConfigurationBuilder =
    this.modify(_.build.runtimeConfig)(_ ++ Map(conf))

  def addRuntimeConfig(key: String, value: String): DataCatererConfigurationBuilder =
    addRuntimeConfig(key -> value)


  def connectionConfig(connectionConfigByName: Map[String, Map[String, String]]): DataCatererConfigurationBuilder =
    this.modify(_.build.connectionConfigByName)(_ ++ connectionConfigByName)

  def connectionConfig(connectionConfigByName: java.util.Map[String, java.util.Map[String, String]]): DataCatererConfigurationBuilder = {
    val scalaConf = toScalaMap(connectionConfigByName)
    val mappedConf = scalaConf.map(c => (c._1, toScalaMap(c._2)))
    connectionConfig(mappedConf)
  }

  def addConnectionConfig(name: String, format: String, connectionConfig: Map[String, String]): DataCatererConfigurationBuilder =
    this.modify(_.build.connectionConfigByName)(_ ++ Map(name -> (connectionConfig ++ Map(FORMAT -> format))))

  def addConnectionConfigJava(name: String, format: String, connectionConfig: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    addConnectionConfig(name, format, toScalaMap(connectionConfig))

  def addConnectionConfig(name: String, format: String, path: String, connectionConfig: Map[String, String]): DataCatererConfigurationBuilder = {
    val pathConf = if (path.nonEmpty) Map(PATH -> path) else Map()
    this.modify(_.build.connectionConfigByName)(_ ++ Map(name -> (connectionConfig ++ Map(FORMAT -> format) ++ pathConf)))
  }

  def addConnectionConfigJava(name: String, format: String, path: String, connectionConfig: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    addConnectionConfig(name, format, path, toScalaMap(connectionConfig))

  def csv(name: String, path: String = "", options: Map[String, String] = Map()): DataCatererConfigurationBuilder =
    addConnectionConfig(name, CSV, path, options)

  def csv(name: String, path: String, options: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    csv(name, path, toScalaMap(options))

  def parquet(name: String, path: String = "", options: Map[String, String] = Map()): DataCatererConfigurationBuilder =
    addConnectionConfig(name, PARQUET, path, options)

  def parquet(name: String, path: String, options: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    parquet(name, path, toScalaMap(options))

  def orc(name: String, path: String = "", options: Map[String, String] = Map()): DataCatererConfigurationBuilder =
    addConnectionConfig(name, ORC, path, options)

  def orc(name: String, path: String, options: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    orc(name, path, toScalaMap(options))

  def json(name: String, path: String = "", options: Map[String, String] = Map()): DataCatererConfigurationBuilder =
    addConnectionConfig(name, JSON, path, options)

  def json(name: String, path: String, options: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    json(name, path, toScalaMap(options))

  def postgres(
                name: String,
                url: String = DEFAULT_POSTGRES_URL,
                username: String = DEFAULT_POSTGRES_USERNAME,
                password: String = DEFAULT_POSTGRES_PASSWORD,
                options: Map[String, String] = Map()
              ): DataCatererConfigurationBuilder =
    addConnection(name, JDBC, url, username, password, options ++ Map(DRIVER -> POSTGRES_DRIVER))

  def postgres(
                name: String,
                url: String,
                username: String,
                password: String,
                options: java.util.Map[String, String]
              ): DataCatererConfigurationBuilder =
    postgres(name, url, username, password, toScalaMap(options))

  def postgres(
                name: String,
                url: String,
                options: java.util.Map[String, String]
              ): DataCatererConfigurationBuilder =
    postgres(name, url, options = toScalaMap(options))

  def postgres(
                name: String,
                url: String
              ): DataCatererConfigurationBuilder =
    postgres(name, url, DEFAULT_POSTGRES_USERNAME)

  def mysql(
             name: String,
             url: String = DEFAULT_MYSQL_URL,
             username: String = DEFAULT_MYSQL_USERNAME,
             password: String = DEFAULT_MYSQL_PASSWORD,
             options: Map[String, String] = Map()
           ): DataCatererConfigurationBuilder =
    addConnection(name, JDBC, url, username, password, options ++ Map(DRIVER -> MYSQL_DRIVER))

  def mysql(
             name: String,
             url: String,
             username: String,
             password: String,
             options: java.util.Map[String, String]
           ): DataCatererConfigurationBuilder =
    mysql(name, url, username, password, toScalaMap(options))

  def mysql(
             name: String,
             url: String,
             options: java.util.Map[String, String]
           ): DataCatererConfigurationBuilder =
    mysql(name, url, options = toScalaMap(options))

  def mysql(
             name: String,
             url: String
           ): DataCatererConfigurationBuilder =
    mysql(name, url, DEFAULT_MYSQL_USERNAME)

  def cassandra(
                 name: String,
                 url: String = DEFAULT_CASSANDRA_URL,
                 username: String = DEFAULT_CASSANDRA_USERNAME,
                 password: String = DEFAULT_CASSANDRA_PASSWORD,
                 options: Map[String, String] = Map()
               ): DataCatererConfigurationBuilder = {
    val sptUrl = url.split(":")
    assert(sptUrl.size == 2, "url should have format '<host>:<port>'")
    val allOptions = Map(
      "spark.cassandra.connection.host" -> sptUrl.head,
      "spark.cassandra.connection.port" -> sptUrl.last,
      "spark.cassandra.auth.username" -> username,
      "spark.cassandra.auth.password" -> password,
    ) ++ options
    addConnectionConfig(name, CASSANDRA, allOptions)
  }

  def cassandra(
                 name: String,
                 url: String,
                 username: String,
                 password: String,
                 options: java.util.Map[String, String]
               ): DataCatererConfigurationBuilder =
    cassandra(name, url, username, password, toScalaMap(options))

  def cassandra(
                 name: String,
                 url: String,
                 options: java.util.Map[String, String]
               ): DataCatererConfigurationBuilder =
    cassandra(name, url, options = toScalaMap(options))

  def cassandra(
                 name: String,
                 url: String
               ): DataCatererConfigurationBuilder =
    cassandra(name, url, DEFAULT_CASSANDRA_USERNAME)

  def jms(name: String, url: String, username: String, password: String, options: Map[String, String] = Map()): DataCatererConfigurationBuilder =
    addConnection(name, JMS, url, username, password, options)

  def solace(
              name: String,
              url: String = DEFAULT_SOLACE_URL,
              username: String = DEFAULT_SOLACE_USERNAME,
              password: String = DEFAULT_SOLACE_PASSWORD,
              vpnName: String = DEFAULT_SOLACE_VPN_NAME,
              connectionFactory: String = DEFAULT_SOLACE_CONNECTION_FACTORY,
              initialContextFactory: String = DEFAULT_SOLACE_INITIAL_CONTEXT_FACTORY,
              options: Map[String, String] = Map()
            ): DataCatererConfigurationBuilder =
    jms(name, url, username, password, Map(
      JMS_VPN_NAME -> vpnName,
      JMS_CONNECTION_FACTORY -> connectionFactory,
      JMS_INITIAL_CONTEXT_FACTORY -> initialContextFactory,
    ) ++ options)

  def solace(
              name: String,
              url: String,
              username: String,
              password: String,
              vpnName: String,
              connectionFactory: String,
              initialContextFactory: String,
              options: java.util.Map[String, String]
            ): DataCatererConfigurationBuilder =
    solace(name, url, username, password, vpnName, connectionFactory, initialContextFactory, toScalaMap(options))

  def solace(
              name: String,
              url: String,
              username: String,
              password: String,
              vpnName: String
            ): DataCatererConfigurationBuilder =
    solace(name, url, username, password, vpnName, DEFAULT_SOLACE_CONNECTION_FACTORY)

  def solace(
              name: String,
              url: String
            ): DataCatererConfigurationBuilder =
    solace(name, url, DEFAULT_SOLACE_USERNAME)

  def kafka(name: String, url: String = DEFAULT_KAFKA_URL, options: Map[String, String] = Map()): DataCatererConfigurationBuilder = {
    addConnectionConfig(name, KAFKA, Map(
      "kafka.bootstrap.servers" -> url,
    ) ++ options)
  }

  def kafka(name: String, url: String, options: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    kafka(name, url, toScalaMap(options))

  def http(name: String, username: String = "", password: String = "", options: Map[String, String] = Map()): DataCatererConfigurationBuilder = {
    val authOptions = if (username.nonEmpty && password.nonEmpty) Map(USERNAME -> username, PASSWORD -> password) else Map()
    addConnectionConfig(name, HTTP, authOptions ++ options)
  }

  def http(name: String, username: String, password: String, options: java.util.Map[String, String]): DataCatererConfigurationBuilder =
    http(name, username, password, toScalaMap(options))

  private def addConnection(name: String, format: String, url: String, username: String,
                            password: String, options: Map[String, String]): DataCatererConfigurationBuilder = {
    addConnectionConfig(name, format, Map(
      URL -> url,
      USERNAME -> username,
      PASSWORD -> password
    ) ++ options)
  }


  def enableGenerateData(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableGenerateData).setTo(enable)

  def enableCount(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableCount).setTo(enable)

  def enableValidation(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableValidation).setTo(enable)

  def enableFailOnError(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableFailOnError).setTo(enable)

  def enableUniqueCheck(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableUniqueCheck).setTo(enable)

  def enableSaveReports(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableSaveReports).setTo(enable)

  def enableSinkMetadata(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableSinkMetadata).setTo(enable)

  def enableDeleteGeneratedRecords(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableDeleteGeneratedRecords).setTo(enable)

  def enableGeneratePlanAndTasks(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableGeneratePlanAndTasks).setTo(enable)

  def enableRecordTracking(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableRecordTracking).setTo(enable)

  def enableGenerateValidations(enable: Boolean): DataCatererConfigurationBuilder =
    this.modify(_.build.flagsConfig.enableGenerateValidations).setTo(enable)


  def planFilePath(path: String): DataCatererConfigurationBuilder =
    this.modify(_.build.foldersConfig.planFilePath).setTo(path)

  def taskFolderPath(path: String): DataCatererConfigurationBuilder =
    this.modify(_.build.foldersConfig.taskFolderPath).setTo(path)

  def recordTrackingFolderPath(path: String): DataCatererConfigurationBuilder =
    this.modify(_.build.foldersConfig.recordTrackingFolderPath).setTo(path)

  def validationFolderPath(path: String): DataCatererConfigurationBuilder =
    this.modify(_.build.foldersConfig.validationFolderPath).setTo(path)

  def generatedReportsFolderPath(path: String): DataCatererConfigurationBuilder =
    this.modify(_.build.foldersConfig.generatedReportsFolderPath).setTo(path)

  def generatedPlanAndTaskFolderPath(path: String): DataCatererConfigurationBuilder =
    this.modify(_.build.foldersConfig.generatedPlanAndTaskFolderPath).setTo(path)

  def recordTrackingForValidationFolderPath(path: String): DataCatererConfigurationBuilder =
    this.modify(_.build.foldersConfig.recordTrackingForValidationFolderPath).setTo(path)


  def numRecordsFromDataSourceForDataProfiling(numRecords: Int): DataCatererConfigurationBuilder =
    this.modify(_.build.metadataConfig.numRecordsFromDataSource).setTo(numRecords)

  def numRecordsForAnalysisForDataProfiling(numRecords: Int): DataCatererConfigurationBuilder =
    this.modify(_.build.metadataConfig.numRecordsForAnalysis).setTo(numRecords)

  def numGeneratedSamples(numSamples: Int): DataCatererConfigurationBuilder =
    this.modify(_.build.metadataConfig.numGeneratedSamples).setTo(numSamples)

  def oneOfMinCount(minCount: Long): DataCatererConfigurationBuilder =
    this.modify(_.build.metadataConfig.oneOfMinCount).setTo(minCount)

  def oneOfDistinctCountVsCountThreshold(threshold: Double): DataCatererConfigurationBuilder =
    this.modify(_.build.metadataConfig.oneOfDistinctCountVsCountThreshold).setTo(threshold)


  def numRecordsPerBatch(numRecords: Long): DataCatererConfigurationBuilder =
    this.modify(_.build.generationConfig.numRecordsPerBatch).setTo(numRecords)

  def numRecordsPerStep(numRecords: Long): DataCatererConfigurationBuilder =
    this.modify(_.build.generationConfig.numRecordsPerStep).setTo(Some(numRecords))


  def numErrorSampleRecords(numRecords: Int): DataCatererConfigurationBuilder =
    this.modify(_.build.validationConfig.numSampleErrorRecords).setTo(numRecords)
}

final case class ConnectionConfigWithTaskBuilder(
                                                  dataSourceName: String = DEFAULT_DATA_SOURCE_NAME,
                                                  options: Map[String, String] = Map()
                                                ) {
  def this() = this(DEFAULT_DATA_SOURCE_NAME, Map())

  def file(name: String, format: String, path: String = "", options: Map[String, String] = Map()): FileBuilder = {
    val configBuilder = DataCatererConfigurationBuilder()
    val fileConnectionConfig = format match {
      case CSV => configBuilder.csv(name, path, options)
      case JSON => configBuilder.json(name, path, options)
      case ORC => configBuilder.orc(name, path, options)
      case PARQUET => configBuilder.parquet(name, path, options)
    }
    setConnectionConfig(name, fileConnectionConfig, FileBuilder())
  }

  def postgres(
                name: String,
                url: String,
                username: String,
                password: String,
                options: Map[String, String] = Map()
              ): PostgresBuilder = {
    val configBuilder = DataCatererConfigurationBuilder().postgres(name, url, username, password, options)
    setConnectionConfig(name, configBuilder, PostgresBuilder())
  }

  def mysql(
             name: String,
             url: String,
             username: String,
             password: String,
             options: Map[String, String] = Map()
           ): MySqlBuilder = {
    val configBuilder = DataCatererConfigurationBuilder().mysql(name, url, username, password, options)
    setConnectionConfig(name, configBuilder, MySqlBuilder())
  }

  def cassandra(
                 name: String,
                 url: String,
                 username: String,
                 password: String,
                 options: Map[String, String] = Map()
               ): CassandraBuilder = {
    val configBuilder = DataCatererConfigurationBuilder().cassandra(name, url, username, password, options)
    setConnectionConfig(name, configBuilder, CassandraBuilder())
  }

  def solace(
              name: String,
              url: String,
              username: String,
              password: String,
              vpnName: String,
              connectionFactory: String,
              initialContextFactory: String,
              options: Map[String, String] = Map()
            ): SolaceBuilder = {
    val configBuilder = DataCatererConfigurationBuilder().solace(name, url, username, password, vpnName, connectionFactory, initialContextFactory, options)
    setConnectionConfig(name, configBuilder, SolaceBuilder())
  }

  def kafka(name: String, url: String, options: Map[String, String] = Map()): KafkaBuilder = {
    val configBuilder = DataCatererConfigurationBuilder().kafka(name, url, options)
    setConnectionConfig(name, configBuilder, KafkaBuilder())
  }

  def http(name: String, username: String, password: String, options: Map[String, String] = Map()): HttpBuilder = {
    val configBuilder = DataCatererConfigurationBuilder().http(name, username, password, options)
    setConnectionConfig(name, configBuilder, HttpBuilder())
  }

  def options(options: Map[String, String]): ConnectionConfigWithTaskBuilder = {
    this.modify(_.options)(_ ++ options)
  }

  def metadataSource(metadataSourceBuilder: MetadataSourceBuilder): ConnectionConfigWithTaskBuilder = {
    this.modify(_.options)(_ ++ metadataSourceBuilder.metadataSource.allOptions)
  }

  private def setConnectionConfig[T <: ConnectionTaskBuilder[_]](name: String, configBuilder: DataCatererConfigurationBuilder, connectionBuilder: T): T = {
    val modifiedConnectionConfig = this.modify(_.dataSourceName).setTo(name)
      .modify(_.options).setTo(configBuilder.build.connectionConfigByName(name))
    connectionBuilder.connectionConfigWithTaskBuilder = modifiedConnectionConfig
    connectionBuilder
  }
}

File: api-java/src/main/java/com/github/pflooky/datacaterer/api/java/model/config/DataCatererConfigurationBuilder.java
package com.github.pflooky.datacaterer.api.java.model.config;

import com.github.pflooky.datacaterer.api.model.Constants;
import com.github.pflooky.datacaterer.api.model.DataCatererConfiguration;
import com.github.pflooky.datacaterer.api.model.FlagsConfig;
import com.github.pflooky.datacaterer.api.model.FoldersConfig;
import com.github.pflooky.datacaterer.api.model.GenerationConfig;
import com.github.pflooky.datacaterer.api.model.MetadataConfig;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.github.pflooky.datacaterer.api.converter.Converters.toScalaMap;
import static com.github.pflooky.datacaterer.api.converter.Converters.toScalaOption;
import static com.github.pflooky.datacaterer.api.converter.Converters.toScalaTuple;

public final class DataCatererConfigurationBuilder {

    private final com.github.pflooky.datacaterer.api.DataCatererConfigurationBuilder scalaDef;

    public DataCatererConfigurationBuilder(com.github.pflooky.datacaterer.api.DataCatererConfigurationBuilder scalaDef) {
        this.scalaDef = scalaDef;
    }

    public com.github.pflooky.datacaterer.api.DataCatererConfigurationBuilder configuration() {
        return scalaDef;
    }

    public DataCatererConfigurationBuilder() {
        this.scalaDef = new com.github.pflooky.datacaterer.api.DataCatererConfigurationBuilder(
                new DataCatererConfiguration(
                        new FlagsConfig(
                                Constants.DEFAULT_ENABLE_COUNT(),
                                Constants.DEFAULT_ENABLE_GENERATE_DATA(),
                                Constants.DEFAULT_ENABLE_RECORD_TRACKING(),
                                Constants.DEFAULT_ENABLE_DELETE_GENERATED_RECORDS(),
                                Constants.DEFAULT_ENABLE_GENERATE_PLAN_AND_TASKS(),
                                Constants.DEFAULT_ENABLE_FAIL_ON_ERROR(),
                                Constants.DEFAULT_ENABLE_UNIQUE_CHECK(),
                                Constants.DEFAULT_ENABLE_SINK_METADATA(),
                                Constants.DEFAULT_ENABLE_SAVE_REPORTS(),
                                Constants.DEFAULT_ENABLE_VALIDATION()
                        ),
                        new FoldersConfig(
                                Constants.DEFAULT_PLAN_FILE_PATH(),
                                Constants.DEFAULT_TASK_FOLDER_PATH(),
                                Constants.DEFAULT_GENERATED_PLAN_AND_TASK_FOLDER_PATH(),
                                Constants.DEFAULT_GENERATED_REPORTS_FOLDER_PATH(),
                                Constants.DEFAULT_RECORD_TRACKING_FOLDER_PATH(),
                                Constants.DEFAULT_VALIDATION_FOLDER_PATH()
                        ),
                        new MetadataConfig(
                                Constants.DEFAULT_NUM_RECORD_FROM_DATA_SOURCE(),
                                Constants.DEFAULT_NUM_RECORD_FOR_ANALYSIS(),
                                Constants.DEFAULT_ONE_OF_DISTINCT_COUNT_VS_COUNT_THRESHOLD(),
                                Constants.DEFAULT_ONE_OF_MIN_COUNT(),
                                Constants.DEFAULT_NUM_GENERATED_SAMPLES()
                        ),
                        new GenerationConfig(
                                Constants.DEFAULT_NUM_RECORDS_PER_BATCH(),
                                toScalaOption(Optional.empty())
                        ),
                        toScalaMap(Collections.emptyMap()),
                        toScalaMap(Collections.emptyMap()),
                        Constants.DEFAULT_SPARK_MASTER()
                )
        );
    }

    public DataCatererConfigurationBuilder sparkMaster(String master) {
        return new DataCatererConfigurationBuilder(scalaDef.sparkMaster(master));
    }

    public DataCatererConfigurationBuilder sparkConfig(Map<String, String> config) {
        return new DataCatererConfigurationBuilder(scalaDef.sparkConfig(toScalaMap(config)));
    }

    public DataCatererConfigurationBuilder addSparkConfig(String key, String value) {
        return new DataCatererConfigurationBuilder(scalaDef.addSparkConfig(toScalaTuple(key, value)));
    }

    private DataCatererConfigurationBuilder connectionConfig(Map<String, Map<String, String>> connectionConfigByName) {
        var mappedConfigs = connectionConfigByName.entrySet().stream()
                .map(c -> Map.entry(c.getKey(), toScalaMap(c.getValue())))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
        return new DataCatererConfigurationBuilder(scalaDef.connectionConfig(toScalaMap(mappedConfigs)));
    }

    private DataCatererConfigurationBuilder addConnectionConfig(String name, String format, Map<String, String> connectionConfig) {
        return new DataCatererConfigurationBuilder(scalaDef.addConnectionConfig(name, format, toScalaMap(connectionConfig)));
    }

    private DataCatererConfigurationBuilder addConnectionConfig(String name, String format, String path, Map<String, String> connectionConfig) {
        return new DataCatererConfigurationBuilder(scalaDef.addConnectionConfig(name, format, path, toScalaMap(connectionConfig)));
    }

    public DataCatererConfigurationBuilder csv(String name, String path, Map<String, String> options) {
        return new DataCatererConfigurationBuilder(scalaDef.csv(name, path, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder csv(String name, String path) {
        return csv(name, path, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder parquet(String name, String path, Map<String, String> options) {
        return new DataCatererConfigurationBuilder(scalaDef.parquet(name, path, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder parquet(String name, String path) {
        return parquet(name, path, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder orc(String name, String path, Map<String, String> options) {
        return new DataCatererConfigurationBuilder(scalaDef.orc(name, path, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder orc(String name, String path) {
        return orc(name, path, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder json(String name, String path, Map<String, String> options) {
        return new DataCatererConfigurationBuilder(scalaDef.json(name, path, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder json(String name, String path) {
        return json(name, path, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder postgres(
            String name,
            String url,
            String username,
            String password,
            Map<String, String> options
    ) {
        return new DataCatererConfigurationBuilder(scalaDef.postgres(name, url, username, password, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder postgres(
            String name,
            String url,
            String username,
            String password
    ) {
        return postgres(name, url, username, password, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder postgres(
            String name,
            String url
    ) {
        return postgres(name, url, "postgres", "postgres");
    }

    public DataCatererConfigurationBuilder mysql(
            String name,
            String url,
            String username,
            String password,
            Map<String, String> options
    ) {
        return new DataCatererConfigurationBuilder(scalaDef.mysql(name, url, username, password, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder mysql(
            String name,
            String url,
            String username,
            String password
    ) {
        return mysql(name, url, username, password, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder mysql(String name, String url) {
        return mysql(name, url, "root", "root");
    }

    public DataCatererConfigurationBuilder cassandra(
            String name,
            String url,
            String username,
            String password,
            Map<String, String> options
    ) {
        return new DataCatererConfigurationBuilder(scalaDef.cassandra(name, url, username, password, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder cassandra(
            String name,
            String url,
            String username,
            String password
    ) {
        return cassandra(name, url, username, password, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder cassandra(String name, String url) {
        return cassandra(name, url, "cassandra", "cassandra");
    }

    public DataCatererConfigurationBuilder jms(String name, String url, String username, String password, Map<String, String> options) {
        return new DataCatererConfigurationBuilder(scalaDef.jms(name, url, username, password, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder solace(
            String name,
            String url,
            String username,
            String password,
            String vpnName,
            String connectionFactory,
            String initialContextFactory,
            Map<String, String> options
    ) {
        return new DataCatererConfigurationBuilder(scalaDef.solace(
                name,
                url,
                username,
                password,
                vpnName,
                connectionFactory,
                initialContextFactory,
                toScalaMap(options))
        );
    }

    public DataCatererConfigurationBuilder solace(
            String name,
            String url,
            String username,
            String password,
            String vpnName
    ) {
        return solace(
                name,
                url,
                username,
                password,
                vpnName,
                Constants.DEFAULT_SOLACE_CONNECTION_FACTORY(),
                Constants.DEFAULT_SOLACE_INITIAL_CONTEXT_FACTORY(),
                Collections.emptyMap()
        );
    }

    public DataCatererConfigurationBuilder solace(String name, String url) {
        return solace(
                name,
                url,
                "admin",
                "admin",
                "default"
        );
    }

    public DataCatererConfigurationBuilder kafka(String name, String url, Map<String, String> options) {
        return new DataCatererConfigurationBuilder(scalaDef.kafka(name, url, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder kafka(String name, String url) {
        return kafka(name, url, Collections.emptyMap());
    }

    public DataCatererConfigurationBuilder http(
            String name,
            String username,
            String password,
            Map<String, String> options
    ) {
        return new DataCatererConfigurationBuilder(scalaDef.http(name, username, password, toScalaMap(options)));
    }

    public DataCatererConfigurationBuilder http(String name) {
        return http(name, "", "", Collections.emptyMap());
    }


    public DataCatererConfigurationBuilder enableGenerateData(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableGenerateData(enable));
    }

    public DataCatererConfigurationBuilder enableCount(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableCount(enable));
    }

    public DataCatererConfigurationBuilder enableValidation(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableValidation(enable));
    }

    public DataCatererConfigurationBuilder enableFailOnError(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableFailOnError(enable));
    }

    public DataCatererConfigurationBuilder enableUniqueCheck(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableUniqueCheck(enable));
    }

    public DataCatererConfigurationBuilder enableSaveReports(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableSaveReports(enable));
    }

    public DataCatererConfigurationBuilder enableSinkMetadata(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableSinkMetadata(enable));
    }

    public DataCatererConfigurationBuilder enableDeleteGeneratedRecords(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableDeleteGeneratedRecords(enable));
    }

    public DataCatererConfigurationBuilder enableGeneratePlanAndTasks(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableGeneratePlanAndTasks(enable));
    }

    public DataCatererConfigurationBuilder enableRecordTracking(boolean enable) {
        return new DataCatererConfigurationBuilder(scalaDef.enableRecordTracking(enable));
    }


    public DataCatererConfigurationBuilder planFilePath(String path) {
        return new DataCatererConfigurationBuilder(scalaDef.planFilePath(path));
    }

    public DataCatererConfigurationBuilder taskFolderPath(String path) {
        return new DataCatererConfigurationBuilder(scalaDef.taskFolderPath(path));
    }

    public DataCatererConfigurationBuilder recordTrackingFolderPath(String path) {
        return new DataCatererConfigurationBuilder(scalaDef.recordTrackingFolderPath(path));
    }

    public DataCatererConfigurationBuilder validationFolderPath(String path) {
        return new DataCatererConfigurationBuilder(scalaDef.validationFolderPath(path));
    }

    public DataCatererConfigurationBuilder generatedReportsFolderPath(String path) {
        return new DataCatererConfigurationBuilder(scalaDef.generatedReportsFolderPath(path));
    }

    public DataCatererConfigurationBuilder generatedPlanAndTaskFolderPath(String path) {
        return new DataCatererConfigurationBuilder(scalaDef.generatedPlanAndTaskFolderPath(path));
    }


    public DataCatererConfigurationBuilder numRecordsFromDataSourceForDataProfiling(int numRecords) {
        return new DataCatererConfigurationBuilder(scalaDef.numRecordsFromDataSourceForDataProfiling(numRecords));
    }

    public DataCatererConfigurationBuilder numRecordsForAnalysisForDataProfiling(int numRecords) {
        return new DataCatererConfigurationBuilder(scalaDef.numRecordsForAnalysisForDataProfiling(numRecords));
    }

    public DataCatererConfigurationBuilder numGeneratedSamples(int numSamples) {
        return new DataCatererConfigurationBuilder(scalaDef.numGeneratedSamples(numSamples));
    }

    public DataCatererConfigurationBuilder oneOfMinCount(long minCount) {
        return new DataCatererConfigurationBuilder(scalaDef.oneOfMinCount(minCount));
    }

    public DataCatererConfigurationBuilder oneOfDistinctCountVsCountThreshold(double threshold) {
        return new DataCatererConfigurationBuilder(scalaDef.oneOfDistinctCountVsCountThreshold(threshold));
    }


    public DataCatererConfigurationBuilder numRecordsPerBatch(long numRecords) {
        return new DataCatererConfigurationBuilder(scalaDef.numRecordsPerBatch(numRecords));
    }

    public DataCatererConfigurationBuilder numRecordsPerStep(long numRecords) {
        return new DataCatererConfigurationBuilder(scalaDef.numRecordsPerStep(numRecords));
    }
}
